File: app/csharp/AccountingSystem.Infrastructure/DAO/Account.cs
using System.ComponentModel.DataAnnotations;

namespace AccountingSystem.Infrastructure.DAO
{
    /// <summary>
    /// 勘定科目マスタ
    /// </summary>
    public class Account
    {
        /// <summary>
        /// 勘定科目ID（主キー）
        /// </summary>
        public int AccountId { get; set; }

        /// <summary>
        /// 勘定科目コード（例：1000, 2000）
        /// </summary>
        [Required(ErrorMessage = "勘定科目コードは必須です")]
        [StringLength(20, ErrorMessage = "勘定科目コードは20文字以内で入力してください")]
        public string AccountCode { get; set; } = string.Empty;

        /// <summary>
        /// 勘定科目名（例：現金、売掛金）
        /// </summary>
        [Required(ErrorMessage = "勘定科目名は必須です")]
        [StringLength(100, ErrorMessage = "勘定科目名は100文字以内で入力してください")]
        public string AccountName { get; set; } = string.Empty;

        /// <summary>
        /// 勘定科目種別（資産、負債、純資産、収益、費用）
        /// </summary>
        [Required(ErrorMessage = "勘定科目種別は必須です")]
        public AccountType AccountType { get; set; }

        /// <summary>
        /// 残高
        /// </summary>
        public decimal Balance { get; set; }

        /// <summary>
        /// BSPL区分（B:貸借対照表, P:損益計算書）
        /// </summary>
        [StringLength(1)]
        public string? BsplCategory { get; set; }

        /// <summary>
        /// 取引要素区分（1:資産, 2:負債, 3:純資産, 4:収益, 5:費用）
        /// </summary>
        [StringLength(1)]
        public string? TransactionElementCategory { get; set; }

        /// <summary>
        /// 費用区分（1:売上原価, 2:販売費及び一般管理費, 3:営業外費用）
        /// </summary>
        [StringLength(1)]
        public string? ExpenseCategory { get; set; }

        /// <summary>
        /// 合計科目（true: 集計科目, false: 明細科目）
        /// </summary>
        public bool IsSummaryAccount { get; set; }

        /// <summary>
        /// 表示順序（財務諸表での表示順）
        /// </summary>
        public int? DisplayOrder { get; set; }

        /// <summary>
        /// 集計対象（true: 集計対象, false: 集計対象外）
        /// </summary>
        public bool IsAggregationTarget { get; set; } = true;

        /// <summary>
        /// 勘定科目カナ（検索用）
        /// </summary>
        [StringLength(40)]
        public string? AccountNameKana { get; set; }

        /// <summary>
        /// 作成日時
        /// </summary>
        public DateTime CreatedAt { get; set; }

        /// <summary>
        /// 更新日時
        /// </summary>
        public DateTime UpdatedAt { get; set; }
    }

    /// <summary>
    /// 勘定科目種別
    /// </summary>
    public enum AccountType
    {
        /// <summary>資産</summary>
        Asset,
        /// <summary>負債</summary>
        Liability,
        /// <summary>純資産</summary>
        Equity,
        /// <summary>収益</summary>
        Revenue,
        /// <summary>費用</summary>
        Expense
    }
}

File: app/scala-ms/financial-accounting-service/src/main/scala/com/example/financial/domain/Account.scala
package com.example.financial.domain

import java.time.LocalDateTime
import com.example.common.domain.*

/**
 * 勘定科目エンティティ
 *
 * 財務会計コンテキストにおける勘定科目を表現します。
 * 勘定科目は取引を分類するための科目であり、
 * 複式簿記の基本的な構成要素です。
 */
case class Account(
    accountId: Option[Long],
    accountCode: AccountCode,
    accountName: String,
    accountType: AccountType,
    isSummaryAccount: Boolean,
    displayOrder: Int,
    isAggregationTarget: Boolean,
    balance: Money,
    createdAt: LocalDateTime,
    updatedAt: LocalDateTime
)

object Account:
  /**
   * 新規勘定科目を作成
   */
  def create(
      accountCode: AccountCode,
      accountName: String,
      accountType: AccountType,
      isSummaryAccount: Boolean = false,
      displayOrder: Int = 0,
      isAggregationTarget: Boolean = true
  ): Account =
    val now = LocalDateTime.now()
    Account(
      accountId = None,
      accountCode = accountCode,
      accountName = accountName,
      accountType = accountType,
      isSummaryAccount = isSummaryAccount,
      displayOrder = displayOrder,
      isAggregationTarget = isAggregationTarget,
      balance = Money.Zero,
      createdAt = now,
      updatedAt = now
    )

  /**
   * 勘定科目の残高を更新
   */
  extension (account: Account)
    def updateBalance(newBalance: Money): Account =
      account.copy(
        balance = newBalance,
        updatedAt = LocalDateTime.now()
      )

    def addDebit(amount: Money): Account =
      account.accountType match
        case AccountType.Asset | AccountType.Expense =>
          account.updateBalance(account.balance + amount)
        case _ =>
          account.updateBalance(account.balance - amount)

    def addCredit(amount: Money): Account =
      account.accountType match
        case AccountType.Liability | AccountType.Equity | AccountType.Revenue =>
          account.updateBalance(account.balance + amount)
        case _ =>
          account.updateBalance(account.balance - amount)
